src/Aspire.Hosting/ResourceBuilderExtensions.cs

```diff
@@ -706,6 +709,123 @@ public static IResourceBuilder<T> WithHealthCheck<T>(this IResourceBuilder<T> bu
         return builder;
     }
 
+    /// <summary>
+    /// Adds a health check to the resource which is mapped to a specific endpoint.
+    /// </summary>
+    /// <typeparam name="T">A resource type that implements <see cref="IResourceWithEndpoints" />.</typeparam>
+    /// <param name="builder">A resource builder.</param>
+    /// <param name="path">The relative path to test.</param>
+    /// <param name="statusCode">The result code to interpret as healthy.</param>
+    /// <param name="endpointName">The name of the endpoint to derive the base address from.</param>
+    /// <returns>A resource builder.</returns>
+    /// <remarks>
+    /// <para>
+    /// This method adds a health check to the health check service which polls the specified endpoint on the resource
+    /// on a periodic basis. The base address is dynamically determined based on the endpoint that was selected. By
+    /// default the path is set to "/" and the status code is set to 200.
+    /// </para>
+    /// </remarks>
+    /// <example>
+    /// This example shows add a HTTP health check to a backend project
+    /// to make sure that the front end does not start until the backend is
+    /// reporting a healthy status based on the return code returned from the
+    /// "/health" path on the backend server.
+    /// <code lang="C#">
+    /// var builder = DistributedApplication.CreateBuilder(args);
+    /// var backend = builder.AddProject&lt;Projects.Backend&gt;("backend")
+    ///                      .WithHttpHealthCheck("/health");
+    /// builder.AddProject&lt;Projects.Backend&gt;("backend")
+    ///        .WithReference(backend).WaitFor(backend);
+    /// </code>
+    /// </example>
+    public static IResourceBuilder<T> WithHttpHealthCheck<T>(this IResourceBuilder<T> builder, string? path = null, int? statusCode = null, string? endpointName = null) where T : IResourceWithEndpoints
+    {
+        path = path ?? "/";
+        endpointName = endpointName ?? "http";
+        statusCode = statusCode ?? 200;
+
+        if (builder.Resource.GetEndpoint(endpointName) is not { Scheme: "http" })
+        {
+            throw new DistributedApplicationException($"The endpoint '{endpointName}' on resource '{builder.Resource.Name}' was not using the 'http' scheme.");
+        }
```

Comment:
I think this check should be deferred, so you can just call it later and it'll fail when it needs to be resolved.